core/src/main/java/org/apache/iceberg/avro/ValueWriters.java

```diff
@@ -137,143 +156,143 @@ private NullWriter() {
     public void write(Void ignored, Encoder encoder) throws IOException {
       encoder.writeNull();
     }
-  }
 
-  private static class BooleanWriter implements ValueWriter<Boolean> {
-    private static final BooleanWriter INSTANCE = new BooleanWriter();
+    @Override
+    public Stream<FieldMetrics> metrics() {
+      return Stream.empty();
+    }
+  }
 
-    private BooleanWriter() {
+  private static class BooleanWriter extends ComparableWriter<Boolean> {
+    private BooleanWriter(int id) {
+      super(id);
     }
 
     @Override
-    public void write(Boolean bool, Encoder encoder) throws IOException {
+    protected void writeVal(Boolean bool, Encoder encoder) throws IOException {
       encoder.writeBoolean(bool);
     }
   }
 
-  private static class ByteToIntegerWriter implements ValueWriter<Byte> {
-    private static final ByteToIntegerWriter INSTANCE = new ByteToIntegerWriter();
-
-    private ByteToIntegerWriter() {
+  private static class ByteToIntegerWriter extends StoredAsIntWriter<Byte> {
+    private ByteToIntegerWriter(int id) {
+      super(id);
     }
 
     @Override
-    public void write(Byte b, Encoder encoder) throws IOException {
-      encoder.writeInt(b.intValue());
+    protected int convert(Byte from) {
+      return from.intValue();
     }
   }
 
-  private static class ShortToIntegerWriter implements ValueWriter<Short> {
-    private static final ShortToIntegerWriter INSTANCE = new ShortToIntegerWriter();
-
-    private ShortToIntegerWriter() {
+  private static class ShortToIntegerWriter extends StoredAsIntWriter<Short> {
+    private ShortToIntegerWriter(int id) {
+      super(id);
     }
 
     @Override
-    public void write(Short s, Encoder encoder) throws IOException {
-      encoder.writeInt(s.intValue());
+    protected int convert(Short from) {
+      return from.intValue();
     }
   }
 
-  private static class IntegerWriter implements ValueWriter<Integer> {
-    private static final IntegerWriter INSTANCE = new IntegerWriter();
-
-    private IntegerWriter() {
+  private static class IntegerWriter extends ComparableWriter<Integer> {
+    private IntegerWriter(int id) {
+      super(id);
     }
 
     @Override
-    public void write(Integer i, Encoder encoder) throws IOException {
+    protected void writeVal(Integer i, Encoder encoder) throws IOException {
       encoder.writeInt(i);
     }
   }
 
-  private static class LongWriter implements ValueWriter<Long> {
-    private static final LongWriter INSTANCE = new LongWriter();
-
-    private LongWriter() {
+  private static class LongWriter extends ComparableWriter<Long> {
```

Comment:
I would expect `LongWriter` to extend `StoredAsLongWriter` instead of `ComparableWriter`. Same for `IntegerWriter`. Could you update those?